techmatters/hrm
# CSAM IWF reporting from Aselo: backend

Jira: https://bugs.benetech.org/browse/CHI-705

This PR is related to:
- https://github.com/techmatters/serverless/pull/142
- https://github.com/techmatters/flex-plugins/pull/543

This PR:
- Adds new migration to create `CSAMReports` table. This table will be used to store the CSAM reports performed against an external service (like IWF). This table is composed by:
  - `id`: unique identifier.
  - `accountSid`: the account this report belongs to.
  - `twilioWorkerId`: the worker who created this report
  - `csamReportId`: the report id sent back from the external source.
  - `contactId`: the contact id related to this report (foreign key). This will initially be null in most cases (CSAM reports are created prior to the creation of the contact record).
  - `createdAt`: time of creation (timestamp).
  - `updatedAt`: last updated (timestamp).
- Adds Sequelize model to work with above table.
- Adds controllers for `CSAMReports`. So far only 4 methods are exposed:
  - `getCSAMReports`: list the CSAM Reports for an account and maybe filtering by contact id.
  - `getCSAMReport` get a single CSAM Report record by id.
  - `createCSAMReport`: creates a new CSAM Report record.
  - `connectToContact`: given a contact id and CSAM Report id, links the two (updating `contactId` column of the report record).
- Modifies contact controller: 
  - Includes the CSAM Report association for contacts.
  - When a new contact record is created, if the request body contains a `csamReports` array of reports, it will connect them to the new contact.
- Exposes `/csamReports` endpoint (only `POST` method so far).

Questions:
- Am I missing some potentially useful column on `CSAMReports`?
- Am I missing some permission or security concern?
- Do we need any other HTTP method to be exposed in this iteration?
- (UI related but impacts here) If the user reloads the UI after submitting a CSAM report but before saving the contact, the link between this two is lost. Is this a concern? Should we start thinking on how to mitigate this?

How to test:
- Run this branch locally (`npm start`).
- Expose the localhost endpoint (using ngrok is the simplest option: `ngrok http <PORT>`).
- Run the above linked Flex Plugin branch (you will need the updated UI).
- Open the browser console and update the following values: 
  - `Twilio.Flex.Manager.getInstance().serviceConfiguration.attributes.hrm_base_url` = `<the https url provided by the ngrok tunel>`
  - `Twilio.Flex.Manager.getInstance().serviceConfiguration.attributes.serverless_base_url` = `<dev environment for serverless>`
- Save a new csam report, save a contact and confirm that those two are stored and linked. 

~WIP:~
- ~CI~
- ~Unit tests~
